### PR TITLE
Fix path type detection for differing remote path type.

### PR DESCRIPTION
**Pull request by brownts**
This change will use the CWD to detect the path type where the
debugger is running and use that to determine whether the target was
supplied with an absolute or relative path.

### CODE REVIEW

**Comment by GitMensch** [comment]
Looks like a reasonable approach.
It is based on the following

> Since the CWD is expected to be an absolute path in the debugger's environment

but that assumption is never checked, is it? (I know that I've used a relative cwd and/or `.` in the past, possibly only for the hack "tell the extension to attach, then do the actual attaching via autorun commands", which doesn't need to work forever.

File: src/backend/mi2/mi2.ts [hunk 1]
```diff
@@ -5,9 +5,7 @@ import { parseMI, MINode } from '../mi_parse';
 import * as linuxTerm from '../linux/console';
 import * as net from "net";
 import * as fs from "fs";
-import { posix } from "path";
-import * as nativePath from "path";
-const path = posix;
+import * as path from "path";
```
Comment (WebFreak001) [inline src/backend/mi2/mi2.ts:8]: why did you change the definition of path? It's used for SSH path building to always use the posix path.

Comment (GitMensch) [reply src/backend/mi2/mi2.ts:8]: That was actually a bad hack. Using the explicit `path.posix` in every code below is definitely better.

Comment (brownts) [reply src/backend/mi2/mi2.ts:8]: Hi @WebFreak001, @GitMensch and I had a discussion on 43cc39e7773f1122ab61cc863256d2264a865e3c where he ran into an issue when the SSH connection was to a Windows machine that the paths were expected to be in Win32 path formatting.  This change was an attempt to "detect" the path type based on the CWD and use the corresponding path manipulation APIs based on that detection.

**Comment by brownts** [comment]
> Looks like a reasonable approach. It is based on the following
> 
> > Since the CWD is expected to be an absolute path in the debugger's environment
> 
> but that assumption is never checked, is it? (I know that I've used a relative cwd and/or `.` in the past, possibly only for the hack "tell the extension to attach, then do the actual attaching via autorun commands", which doesn't need to work forever.

@GitMensch, I was trying to think of a case in which a non-absolute CWD might be supplied and we'd think it was okay to do so.  Your scenario might be the case...although a probably unexpected case.  However, with an attach, we don't use `target` within `initCommands` anyway, so even if it was constructed with the wrong path type, that is probably fine (since it isn't subsequently used).  I just tested the attach with GDB for attaching to a PID as well as a remote attach and they worked fine with a "." CWD, since the subsequent target is not used.

I did notice other places of the code (outside the `initCommands`) that likely don't work with attaching when you don't supply an absolute CWD.   Once place is the LLDB attach (since it issues the `file-exec-and-symbols` command [here](https://github.com/WebFreak001/code-debug/blob/master/src/backend/mi2/mi2lldb.ts#L45)).  Another would likely be the "extended remote".  So your method probably only works in certain attach scenarios but not all, but I don't believe the changes here would have affected those scenarios.  Additionally, I'd think if the changes as described in #329 are implemented, this would probably at least make the GDB attach to PID not work with your method either, if it follows the LLDB approach.

**Comment by brownts** [comment]
If this still doesn't seem sufficient, I could first check that the target path is not a POSIX absolute path and not a Win32 path before attempting to detect the path type from the CWD for the subsequent join.  This might be a little safer in the case where the target is already specified as an absolute path, but the CWD isn't.
```typescript
if (!path.posix.isAbsolute(target) && !path.win32.isAbsolute(target)) {
	const debuggerPath = path.posix.isAbsolute(cwd) ? path.posix : path.win32;
	target = debuggerPath.join(cwd, target);
}
```
I think a Win32 relative path using a leading `/` path separator could be misinterpreted this way as a POSIX absolute path though.  Let me know what you think of this, or if you have a better idea on an approach.

**Comment by brownts** [comment]
Any decision on this?  I'd like to resolve this as HEAD is essentially broken right now for SSH connections due to this problem.  If the fix is to just to do the change as suggested in https://github.com/WebFreak001/code-debug/commit/43cc39e7773f1122ab61cc863256d2264a865e3c, that's fine, but this problem blocking me from addressing some of the other issues.

**Comment by GitMensch** [comment]
_Seems like it did not come through the mail gateway, so here directly again._

I'm undecided.... if @WebFreak001 does not say otherwise then I'll merge tomorrow (= this evening) as I guess you've tested SSH from posix->posix, win32->win32, posix->win32, win32->posix.

If you didn't test some of those please drop a note and I'll test the missing pieces directly after the merge.

**Comment by GitMensch** [comment]
Thanks for working on this.

If you know a test was missing, please drop a note (in general it will be better than my half commit that was there before).

**Comment by brownts** [comment]
I have a setup on Windows now as well.  I did test all 4 combinations that you mention above and they all seem to be working as expected.
